fs: make FSWatcher.start private

lholmquist opened the pull request:
* This is a semver-major change to rename the FSWatcher.start
function to FSWatcher._start to make it private

Also see #29872 

The motivation here is that it serves no purpose to the end user.
An instance of FSWatcher is returned when a user calls fs.watch,
which will call the start method.  A user can't create an instance
of a FSWatcher directly.  If the start method is called by a user
it is a noop since the watcher has already started.  Calling start
after a watcher has closed is also a noop

@addaleax @cjihrig I went straight for changing the function to private, do we need an entry or anything in the deprecations.md?

<!--
Thank you for your pull request. Please provide a description above and review
the requirements below.

Bug fixes and new features should include tests and possibly benchmarks.

Contributors guide: https://github.com/nodejs/node/blob/master/CONTRIBUTING.md
-->

##### Checklist
<!-- Remove items that do not apply. For completed items, change [ ] to [x]. -->

- [x] `make -j4 test` (UNIX), or `vcbuild test` (Windows) passes
- [ ] tests and/or benchmarks are included
- [ ] documentation is changed or added
- [ ] commit message follows [commit guidelines](https://github.com/nodejs/node/blob/master/doc/guides/contributing/pull-requests.md#commit-message-guidelines)

<!--
Developer's Certificate of Origin 1.1

By making a contribution to this project, I certify that:

(a) The contribution was created in whole or in part by me and I
    have the right to submit it under the open source license
    indicated in the file; or

(b) The contribution is based upon previous work that, to the best
    of my knowledge, is covered under an appropriate open source
    license and I have the right under that license to submit that
    work with modifications, whether created in whole or in part
    by me, under the same open source license (unless I am
    permitted to submit under a different license), as indicated
    in the file; or

(c) The contribution was provided directly to me by some other
    person who certified (a), (b) or (c) and I have not modified
    it.

(d) I understand and agree that this project and the contribution
    are public and that a record of the contribution (including all
    personal information I submit with it, including my sign-off) is
    maintained indefinitely and may be redistributed consistent with
    this project or the open source license(s) involved.
-->

lholmquist commented:
@addaleax good to know about the symbols.  I don't suppose you have a link to that usage handy :)

addaleax commented:
@lholmquist There is some info in https://github.com/nodejs/node/blob/master/doc/guides/using-symbols.md – a basic usage example would be:

```js
const kStart = Symbol('kStart');

FSWatcher.prototype[kStart] = function() { ... };

const watcher = new FSWatcher();
watcher[kStart]();
```

addaleax commented:
@lholmquist You could add the symbol to the exports of `lib/internal/fs/watchers.fs` and use that in `lib/fs.js` – it’s an internal module, so it’s still hard to access the symbol from userland code.

lholmquist commented:
@addaleax ok,  i'll do that.  wasn't sure if it was ok to add another export, but like you said, it is an internal module

lholmquist commented:
I added a commit that is using a Symbol.  The reason for naming it the way i did was becuase i think something similar will probably happen with the `StatWatcher.start` method, so i wanted to make sure there was that differentiation 

I also removed the tests for FSWatcher.start

BridgeAR commented:
What's the semverness here?

lholmquist commented:
@BridgeAR semver-major because it is removing a public non-documented api https://github.com/nodejs/node/pull/29872#issuecomment-539027301

Trott commented:
CITGM: https://ci.nodejs.org/view/Node.js-citgm/job/citgm-smoker/2053/

Trott commented:
Added to the 13.0.0 milestone, but if CITGM reveals issues, we can take it off.

Trott commented:
> CITGM: https://ci.nodejs.org/view/Node.js-citgm/job/citgm-smoker/2053/

Results look good.

Trott commented:
Landed in 7eacb74

jasnell commented:
This really should have been a deprecation for the existing API

Trott commented:
> This really should have been a deprecation for the existing API

Not too late to revert and go that route if you feel strongly about it (and no opposition from @addaleax, @cjihrig, @BridgeAR...). This hasn't gone out in a release yet.

jasnell commented:
A new pr that adds a deprecated alias is good.
